jena-arq/src/main/java/org/apache/jena/sparql/core/assembler/DatasetAssembler.java

```diff
@@ -58,27 +64,33 @@ public Dataset createDataset(Assembler a, Resource root, Mode mode) {
             // Assembler description did not define one.
             dftModel = GraphFactory.makeDefaultModel() ;
         Dataset ds = DatasetFactory.create(dftModel) ;
-        // -------- Named graphs
-        List<RDFNode> nodes = GraphUtils.multiValue(root, DatasetAssemblerVocab.pNamedGraph) ;
-        for ( RDFNode n : nodes ) {
-            if ( !(n instanceof Resource) )
-                throw new DatasetAssemblerException(root, "Not a resource: " + FmtUtils.stringForRDFNode(n));
-            Resource r = (Resource)n;
-
-            String gName = GraphUtils.getAsStringValue(r, DatasetAssemblerVocab.pGraphName);
-            Resource g = GraphUtils.getResourceValue(r, DatasetAssemblerVocab.pGraph);
-            if ( g == null ) {
-                g = GraphUtils.getResourceValue(r, DatasetAssemblerVocab.pGraphAlt);
-                if ( g != null ) {
-                    Log.warn(this, "Use of old vocabulary: use :graph not :graphData");
-                } else {
-                    throw new DatasetAssemblerException(root, "no graph for: " + gName);
+        Txn.executeWrite(ds, () -> {
+            // Load data into the default graph or quads into the dataset.
+            multiValueAsString(root, data)
+                .forEach(dataURI -> read(ds, dataURI));
+    
```

Comment:
This interacts with the setting up of the dft graph/named graphs. It's going to add dft graph just made (which might have content via `ja:content) but also the named graphs.  It's a dataset by links to the models where adding a model replaces the model.

I'm not sure what he best balance is here -- options include (1) warning and no action on seeing `ja:data` (with a note to use TIM) (2) do after all the named models are added because parsing is adding quads, not replacing models (3) allow this or specified models, which would be better done with TIM (4) it's `ja:data` or specify graphs, not a combination; if `ja:data` return a TIM.

(4) looks like a good way to encourage TIM usage. If (4) then a plain, empty `ja:RDFDataset` could be TIM as well.

Reply:
I'm trying to think of a use case for "load quads into non-TIM" and the one that occurs to me is in an embedded or integrated situation where you have a lot of quads, like so many that you prefer the memory-parsimonious-ness of the general IM dataset, maybe because you have other processes running in the system. Sound likely enough to merit (2)?

Reply:
That is to load specialised graphs e.g. inference? In which case the graphs themselves need to be pre-created before reading data. We already have a mechanism for graphs with `ja:content` on the graph description. Having two ways seems bad. Strange things will happen if a graph is in the quads and not explicitly created (it will be a plain graph when inference expected). It could be done be splitting the quads into separate graphs and generating the assembler text. 

It might make sense to have some plain graphs and specialist graphs, that is a disjoint set, but to enforce this to stop accidents is not so easy without a lot of additional machinery to enforce disjoint.

How about doing this progressively - we make ja:RDFDataset create either TIM, or linked datasets. And for now, it is one or the other, not a combination. (We can have per graph data loading into TIM as it does at the moment with some care - look into the namedGhraph property to see if it has `ja:data`+name, and only those, if not, it's a general model).

We leave more exotic combinations until we get a request because that may help choose what to do with the corner cases like whether they overlap or not. If we provide now, the corner cases are fixed by compatibility.

Reply:
Okay, if I get what you are saying, it's:

1. Check to see if quads are being loaded, if so, TIM. 
2. Otherwise, check the named graphs. If they are all `ja:data` guys, then TIM again.
3. Otherwise, general dataset.

I'll get on this later today.

Reply:
@afs What's a good idiom for switching to a new assembler? In other words, let's say the code tests for the presence of quads and finds them and it's time to pivot to TIM. Obviously, I could just `new InMemDatasetAssembler()`, but I'm thinking there must be a more elegant way. I looked at `AssemblerUtils` but only saw ways to register `Assembler`s, not retrieve them…